### DISCUSSION TITLE
Crash when using flexcomm5 as i2c on LPC5526

### DISCUSSION THREAD

**Discussion by RomainPelletant** [zephyrproject-rtos/zephyr#52727] [Q&A] [answered]
**Describe the bug**
Division by zero when Zephyr boot if flexcomm5 is configured as i2c on LPC5526

Please also mention any information which could help others to understand
the problem you're facing:
 - LPC5526 with Zephyr 3.1
 - pin muxing is like
 `#if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(flexcomm5), nxp_lpc_i2c, okay) && CONFIG_I2C
  /* I2C5 SDA */
  uint32_t port0_pin8_config = (
    IOCON_FUNC3 |
    IOCON_MODE_PULLUP |
    IOCON_PIO_INV_DI |
    IOCON_PIO_DIGITAL_EN |
    IOCON_PIO_SLEW_STANDARD |
    IOCON_PIO_OPENDRAIN_DI);

  /* I2C5 SCL */
  uint32_t port1_pin15_config = (
    IOCON_FUNC4 |
    IOCON_MODE_PULLUP |
    IOCON_PIO_INV_DI |
    IOCON_PIO_DIGITAL_EN |
    IOCON_PIO_SLEW_STANDARD |
    IOCON_PIO_OPENDRAIN_DI);

  pinmux_pin_set(port0, 8, port0_pin8_config);
  pinmux_pin_set(port1, 15, port1_pin15_config);
#endif`

**To Reproduce**
1. mkdir build; cd build
2. cmake -DBOARD=lpc55S28
3. set flexcomm5 as i2c
4. make
5. See error

**Expected behavior**
I2C starts

**Workaround**
in i2c_mcux_flexcomm.c, in mcux_flexcomm_init function,  _clock_control_get_rate(config->clock_dev, config->clock_subsys, &clock_freq)_ returns 0 (for clock_freq). master_config.baudRate_Bps is at 0.
If clock_freq is manually set (hard coded) at 96000000 and master_config.baudRate_Bps at 100000, everything works well.

**Logs and console output**
`[00:00:00.001,000] <err> os: ***** USAGE FAULT *****
[00:00:00.001,000] <err> os:   Division by zero
[00:00:00.001,000] <err> os: r0/a1:  0x40096000  r1/a2:  0x000186a0  r2/a3:  0x00000000
[00:00:00.001,000] <err> os: r3/a4:  0x00000000 r12/ip:  0x00000000 r14/lr:  0x00000000
[00:00:00.001,000] <err> os:  xpsr:  0x09000000
[00:00:00.001,000] <err> os: s[ 0]:  0x00000000  s[ 1]:  0x00000000  s[ 2]:  0x00000000  s[ 3]:  0x00000000
[00:00:00.001,000] <err> os: s[ 4]:  0x00000000  s[ 5]:  0x00000000  s[ 6]:  0x00000000  s[ 7]:  0x00000000
[00:00:00.001,000] <err> os: s[ 8]:  0x00000000  s[ 9]:  0x00000000  s[10]:  0x00000000  s[11]:  0x00000000
[00:00:00.001,000] <err> os: s[12]:  0x0000`

**Comment by RomainPelletant** [comment]
It appears the application does not crash anymore but i2c is not usable.
The following function returns 0 clock_control_get_rate as frequency.

In fsl_clock.h, CLOCK_GetCTimerClkFreq returns 0. This is why it is crashing. What is the clock frequency for flexcomm5?

**Comment by DerekSnell** [comment, marked as answer]
Hi @RomainPelletant ,

The clock function returns 0 Hz because the clock signal to Flexcomm5 is not enabled.  The clocks are initialized in soc.c.  These [lines in soc.c](https://github.com/zephyrproject-rtos/zephyr/blob/e6ef1f3434a2f7c26bd9d1bbdba8537ff541ef01/soc/arm/nxp_lpc/lpc55xxx/soc.c#L109) are where some other Flexcomm clocks are selected and enabled.  The same needs to be done for Flexcomm5.

One option is to modify this soc.c file.  However, clock configuration will depend on the board/application.  If I were changing this, I would want to control it in my own board file, not in the shared soc.c.  Your board file could add another hardware initialization function similar to [clock_init()](https://github.com/zephyrproject-rtos/zephyr/blob/e6ef1f3434a2f7c26bd9d1bbdba8537ff541ef01/soc/arm/nxp_lpc/lpc55xxx/soc.c#L59).  And you can add your init function to Zephyr's start-up initialization using the [SYS_INIT macro](https://github.com/zephyrproject-rtos/zephyr/blob/e6ef1f3434a2f7c26bd9d1bbdba8537ff541ef01/soc/arm/nxp_lpc/lpc55xxx/soc.c#L261).  Just be aware of the initialization order.  You will need your init function that enables the Flexcomm5 clock to be called before the Flexcomm driver initialization.  The webinar [Application Portability Made Easy With Zephyr OS and NXP](https://www.nxp.com/design/training/application-portability-made-easy-with-zephyr-os-and-nxp:TIP-APPLICATION-PORTABILITY-MADE-EASY-WITH-ZEPHYR) has more details on the start-up process.

Best regards

**Comment by RomainPelletant** [comment]
Thanks @DerekSnell
